### PR TITLE
Add `leave` argument to RichProgressBar

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Add `display_every_n_epochs` argument to `RichProgressBar`. Set to a non-negative integer to display progress bar every n epochs.

Screen Shot 2021-11-02 at 1 27 05 PM

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by tchaton** [comment]
Hey @kaushikb11.

By the same occasion, do you think you could look into the fickering on the progress bar. The last line isn't aligned properly due the loss.

Best,
T.C

File: pytorch_lightning/callbacks/progress/rich_progress.py [hunk 1]
```diff
@@ -195,6 +196,7 @@ class RichProgressBar(ProgressBarBase):
 
     Args:
         refresh_rate_per_second: the number of updates per second. If refresh_rate is 0, progress bar is disabled.
+        display_every_n_epochs: Set to a non-negative integer to display progress bar every n epochs.
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/progress/rich_progress.py:199]: should it be restricted to just -1?

File: pytorch_lightning/callbacks/progress/rich_progress.py [hunk 2]
```diff
@@ -323,9 +331,19 @@ def on_train_epoch_start(self, trainer, pl_module):
         total_batches = total_train_batches + total_val_batches
 
         train_description = self._get_train_description(trainer.current_epoch)
+        if (
+            self.main_progress_bar_id is not None
+            and self._display_every_n_epochs
+            and (not trainer.current_epoch % self._display_every_n_epochs)
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/progress/rich_progress.py:337]: so it's like if I set it to `2` it won't show at the 3rd, 5th, 7th epoch?

Comment (rohitgr7) [inline pytorch_lightning/callbacks/progress/rich_progress.py:337]: also I think current_epoch is zero-indexed, so should it be `(trainer.current_epoch + 1) % self._display_every_n_epochs` ?

File: tests/callbacks/test_rich_progress_bar.py [hunk 3]
```diff
@@ -159,3 +160,29 @@ def configure_columns(self, trainer, pl_module):
...
+        progress_bar = RichProgressBar(display_every_n_epochs=2)
+        trainer = Trainer(
+            default_root_dir=tmpdir,
+            num_sanity_val_steps=0,
+            limit_train_batches=1,
+            max_epochs=6,
+            callbacks=progress_bar,
+        )
+        trainer.fit(model)
+    assert mock_progress_stop.call_count == 3
```
Comment (rohitgr7) [inline tests/callbacks/test_rich_progress_bar.py:182]: should we instead check at what epoch the progress bar comes up?

File: pytorch_lightning/callbacks/progress/rich_progress.py (hunk 1 above)
Comment (kaushikb11) [reply pytorch_lightning/callbacks/progress/rich_progress.py:199]: What do you mean by `restricted to just -1`?

File: pytorch_lightning/callbacks/progress/rich_progress.py (hunk 2 above)
Comment (kaushikb11) [reply pytorch_lightning/callbacks/progress/rich_progress.py:337]: It will show, but it won't start as a new bar

Comment (kaushikb11) [reply pytorch_lightning/callbacks/progress/rich_progress.py:337]: No, I don't believe so. It should go with every 2 epochs starting with zero index

File: pytorch_lightning/callbacks/progress/rich_progress.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/callbacks/progress/rich_progress.py:199]: I meant.. currently setting it to -2, -100 will show it on every epochs. so should we just make it -1? else raise a Misconfig.. the reason being, we have certain other flags across codebase that takes just `-1` as an indication for no restriction.

File: tests/callbacks/test_rich_progress_bar.py (hunk 3 above)
Comment (kaushikb11) [reply tests/callbacks/test_rich_progress_bar.py]: Updated the test! `reset` means continuing on the same bar.

File: pytorch_lightning/callbacks/progress/rich_progress.py (hunk 1 above)
Comment (kaushikb11) [reply pytorch_lightning/callbacks/progress/rich_progress.py:199]: No, it will currently [raise](https://github.com/PyTorchLightning/pytorch-lightning/blob/support%2Fdisplay_every_n_epoch/pytorch_lightning/callbacks/progress/rich_progress.py#L219) a Misconfiguration exception.

Comment (rohitgr7) [reply pytorch_lightning/callbacks/progress/rich_progress.py:199]: okay.. sorry my bad.. it's non-negative.. 🤦‍♂️

**Review by rohitgr7 — APPROVED** [review]

**Comment by kaushikb11** [comment]
> do you think you could look into the fickering on the progress bar.

@tchaton The flickering issue seems to exist for VS code terminal. It works as expected on ITerm2. I think we would need to raise an issue on Rich repo. cc @SeanNaren

**Review by tchaton — APPROVED** [review]
LGMT !

File: pytorch_lightning/callbacks/progress/rich_progress.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/callbacks/progress/rich_progress.py:199]: "display" is misleading as a name imo. I wasn't sure what you mean by it, and I tested it out. It actually means you want to keep the last line of the progress bar in the stdout print at the end of an epoch, if I interpret correctly. 

Have we considered/discussed the name to be something else? Why do you think it is important to set the frequency to a different value than 0 or 1? 

`RichProgressBar(leave=True|False)`
`RichProgressBar(leave_every_epoch=True|False)`

File: CHANGELOG.md [hunk 4]
```diff
@@ -128,6 +128,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added support for empty `gpus` list to run on CPU ([#10246](https://github.com/PyTorchLightning/pytorch-lightning/pull/10246))
 - Added `configure_columns` method to `RichProgressBar` ([#10288](https://github.com/PyTorchLightning/pytorch-lightning/pull/10288))
 - Added a warning if multiple batch sizes are found from ambiguous batch ([#10247](https://github.com/PyTorchLightning/pytorch-lightning/pull/10247))
+- Added `display_every_n_epochs` argument to `RichProgressBar` ([#10301](https://github.com/PyTorchLightning/pytorch-lightning/pull/10301))
```
[the remark this answers is not in the document]
Comment (kaushikb11) [reply CHANGELOG.md]: Updated!

File: pytorch_lightning/callbacks/progress/rich_progress.py (hunk 1 above)
Comment (SeanNaren) [reply pytorch_lightning/callbacks/progress/rich_progress.py:199]: how about `bar_every_epoch`? and set it to true false rather than providing an actual value?

Comment (awaelchli) [reply pytorch_lightning/callbacks/progress/rich_progress.py:199]: But if I understand (hopefully), the question is not about whether the user wants to see a bar or not (it will be there the entire time). The question is whether once it is full, should it delete the full bar and reset completely or should it make a new line to start a new bar (and leave the old one behind).

Comment (kaushikb11) [reply pytorch_lightning/callbacks/progress/rich_progress.py:199]: I am okay with boolean values.

`display_bar_every_epoch` or `bar_every_epoch`?

Tbh, `display_every_epoch` sounds good to me. Because you are technically displaying each epoch progress bar to user on the terminal.

Comment (rohitgr7) [reply pytorch_lightning/callbacks/progress/rich_progress.py:199]: `progress_every_epoch`?

Comment (kaushikb11) [reply pytorch_lightning/callbacks/progress/rich_progress.py:199]: Or should we go with `reset_every_epoch` with True as default.

Comment (carmocca) [reply pytorch_lightning/callbacks/progress/rich_progress.py:199]: I don't have the full context. But one clarification:

The progress bar still appears and gets updated during the epoch and this feature is about whether it stays printed on the terminal afterward?

If that's correct, I'm in favor of naming it `leave` which matches exactly the `tqdm`
 API

Comment (Borda) [reply pytorch_lightning/callbacks/progress/rich_progress.py:199]: so if I read it correctly, keeps showing last n bars from particular epochs, so `preserve_n_last_epochs_bars`?

Comment (justusschock) [reply pytorch_lightning/callbacks/progress/rich_progress.py:199]: > I think we could be more intuitive about it and we don't need to match the tqdm api

I agree, but I think it is easier for people if we provide the same interface for the same functionality from different backends.

Comment (kaushikb11) [reply pytorch_lightning/callbacks/progress/rich_progress.py:199]: @tchaton @SeanNaren I need two more votes, before going ahead with it. Let's make it quick!

**Review by awaelchli — APPROVED** [review]
